Malmo/src/TCPServer.cpp

```diff
@@ -89,17 +100,38 @@ namespace malmo
             boost::asio::placeholders::error));
     }
     
-    void TCPServer::handleAccept( 
+    void TCPServer::handleAccept(
         boost::shared_ptr<TCPConnection> new_connection,
         const boost::system::error_code& error)
     {
         if (!error)
         {
-            new_connection->read();
-            this->startAccept();
+            if (this->closing)
+            {
+                new_connection.get()->getSocket().close();
+                if (scope != nullptr) 
+                    scope->release();
+            }
+            else {
+                new_connection->read();
+                if (!this->closing)
+                {
+                    this->startAccept();
+                }
+                else
+                {
+                    if (scope != nullptr)
+                        scope->release();
+                }
+            }
         }
         else
+        {
             LOGERROR(LT("TCPServer::handleAccept("), this->log_name, LT(") - "), error.message());
+            if (closing && scope != nullptr) {
```

Comment:
I'm not entirely sure what error conditions would cause this code to be executed - maybe things are already foobar by this point - but if `closing` isn't true, `scope` will never get released, so will presumably leak? Not sure how much this matters?

Reply:
I'm sure it's fine, but (unless I've overlooked something) if we never call `scope->release()`, then the owning VideoServer (or StringServer or whatever) never releases its shared_ptr to itself, so there's a cyclic dependency and the VideoServer/StringServer will never get deleted. AgentHost will call `close()` on the VideoServer, and then create a new one, but the old one will still exist, along with its TCPServer, std::ofstream and boost::mutex. Not sure how often such accept errors crop up, maybe there's no danger at all. But if it's something that could be triggered, say, by the user calling `startMission` at the wrong time, in a long-running experiment comprising many thousands of missions, it could end up being a problem...? I think we could just remove the `closing` from the `if (closing && scope != nullptr)` line, so that scope is always released in the error case.

Reply:
Ok, will release on error (removing closing check).